src/Airship/Internal/Helpers.hs

```diff
@@ -94,21 +95,40 @@ toWaiResponse Response{..} cfg trace quip =
                       else []
 
 -- | Given a 'RoutingSpec', a 404 resource, and a user state @s@, construct a WAI 'Application'.
-resourceToWai :: AirshipConfig -> RoutingSpec IO () -> Resource IO -> Wai.Application
-resourceToWai cfg routes resource404 =
-  resourceToWaiT cfg (const id) routes resource404
-
--- | Given a 'RoutingSpec', a 404 resource, and a user state @s@, construct a WAI 'Application'.
-resourceToWaiT :: Monad m => AirshipConfig -> (Request -> m Wai.Response -> IO Wai.Response) -> RoutingSpec m () -> Resource m -> Wai.Application
-resourceToWaiT cfg run routes resource404 req respond = do
+resourceToWai :: AirshipConfig
+              -> RoutingSpec IO ()
+              -> ErrorResponses
+              -> Wai.Application
+resourceToWai cfg routes =
+    resourceToWaiT cfg (const id) routes
+
+-- | Given a 'RoutingSpec', an 'ErrorResponses', and a user state @s@, construct a WAI 'Application'.
+resourceToWaiT :: Monad m =>
+                  AirshipConfig
+               -> (Request -> m Wai.Response -> IO Wai.Response)
+               -> RoutingSpec m ()
+               -> ErrorResponses
+               -> Wai.Application
+resourceToWaiT cfg run routes errors req respond = do
     let routeMapping = runRouter routes
         pInfo = Wai.pathInfo req
-        (resource, (params', matched)) = route routeMapping pInfo resource404
-    nowTime <- getCurrentTime
+        response404 = M.lookup HTTP.status404 errors
     quip <- getQuip
-    (=<<) respond . run req $ do
-      (response, trace) <- eitherResponse nowTime params' matched req (flow resource)
-      return $ toWaiResponse response cfg (traceHeader trace) quip
+    case route routeMapping pInfo of
+        Nothing -> do
+            -- No route matched so bail out
+            let (ct, respBody) = fromMaybe (mempty, Empty) response404
+                resp = Response HTTP.status404 [(HTTP.hContentType, ct)] respBody
+            respond =<< (return $ toWaiResponse resp cfg "" quip)
+        Just (resource, (params', matched)) -> do
+            nowTime <- getCurrentTime
+            -- Merge the application-wide error responses with the resource
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            (=<<) respond . run req $ do
```

Comment:
> This (the prefix-applied bind) is sort of an odd idiom here;

Apologies, that was me, it's a bad habit of mine to avoid coming up with variable names.